Prevent losing items after backpack size is reduced

WheteThunger opened the pull request:
This change addresses an issue where a player can lose items from their backpack if opened after their permissions or the plugin configuration were changed to not allow as much space.

https://umod.org/community/backpacks/20710-delete-bottom-slots-when-permission-revoked.

The compaction logic could be improved for the use case where there are multiple stacks in the player inventory or backpack that could already be combined, but that's unlikely to happen and therefore probably isn't worth the effort.

The dropping logic could also be improved to create a container and drop that if there's more than one item, but this is already a good start.

This does not address the use case where permissions to access the backpack are completely revoked, as the player won't even be able to use the backpack command. In that case, the items in their backpack are inaccessible, but not lost.

WheteThunger committed:
Prevent losing items after backpack size is reduced

When a player opens their own backpack after its size has been reduced, this deals with any extra items by compacting the backpack, attempting to give the items to the player and eventually dropping the rest as a last resort.

LaserHydra commented on Backpacks.cs:
```diff
@@ -651,6 +661,37 @@ public void Open(BasePlayer looter)
 
+                if (looter.userID == OwnerId)
+                {
+                    // Check for items beyond the backpack's capacity
+                    var allowedCapacity = GetCapacity();
+
+                    var extraItems = _itemContainer.itemList
+                        .OrderBy(item => item.position)
+                        .Where(item => item.position >= allowedCapacity)
+                        .ToList();
+
+                    var itemsDroppedOrGivenToPlayer = 0;
+                    foreach (var item in extraItems)
+                    {
+                        // Try to add to a vacant slot or an existing stack in the backpack or player inventory, else drop
+                        item.RemoveFromContainer();
+                        if (!item.MoveToContainer(_itemContainer))
+                        {
+                            itemsDroppedOrGivenToPlayer++;
+                            if (!looter.inventory.GiveItem(item))
+                            {
+                                item.Drop(looter.GetDropPosition(), looter.GetDropVelocity(), looter.GetNetworkRotation());
+                            }
+                        }
+                    }
+
+                    if (itemsDroppedOrGivenToPlayer > 0)
+                    {
+                        _instance.PrintToChat(looter, _instance.lang.GetMessage("Backpack Over Capacity", _instance, looter.UserIDString));
+                    }
```
Please move this to a separate method, like `HandleOverflowingItems(BasePlayer itemReceiver/owner)` :)

LaserHydra replied:
Also please restore the original capacity after moving the overflowing items.

LaserHydra commented on Backpacks.cs:
```diff
@@ -651,6 +661,37 @@ public void Open(BasePlayer looter)
+                    var allowedCapacity = GetCapacity();
+
+                    var extraItems = _itemContainer.itemList
+                        .OrderBy(item => item.position)
```
I believe this OrderBy statement is not required and produces a bit of overhead. Unless you disagree, please remove it.

WheteThunger replied:
Added an inline comment attempting to explain why it's useful. Moved to a different logic branch to improve performance.

LaserHydra commented on Backpacks.cs:
```diff
@@ -582,6 +583,14 @@ public void Initialize()
                     _itemContainer.GiveUID();
 
+                    var allowedCapacity = _itemContainer.capacity;
+                    if (_itemDataCollection.Count() > allowedCapacity)
```
Please use the _itemDataCollection.Count property here, instead of the Linq .Count() method. Thanks :)

LaserHydra commented on Backpacks.cs:
```diff
@@ -582,6 +583,14 @@ public void Initialize()
+                    {
+                        // Temporarily increase the capacity to allow all items to fit
+                        // Extra items will be addressed when the backpack is opened by the owner
+                        _itemContainer.capacity = MaxSize * SlotsPerRow;
```
This should likely be `allowedCapacity = MaxSize * SlotsPerRow;`, otherwise it's just going to be overwritten in line 604 :)

WheteThunger replied:
That's actually on purpose. It turns out that reducing the capacity after adding items does not remove the items over capacity. My original approach was to increase the capacity in this method and then reduce it in the `Open()` method after taking the items out. However, I discovered near the top of this `Initialize()` method that the capacity was being reset so the check for higher capacity in the `Open()` method wasn't working.

We can do it either way I suppose. I guess the only concern with my current approach is that while it works now (items are not lost), who knows if this could change in the future.

I have updated it now to use the original approach. Took some extra work to deal with the edge cases.

LaserHydra commented on Backpacks.cs:
```diff
@@ -582,6 +583,14 @@ public void Initialize()
                     _itemContainer.allowedContents = ItemContainer.ContentsType.Generic;
                     _itemContainer.GiveUID();
 
+                    var allowedCapacity = _itemContainer.capacity;
```
This should probably be `var allowedCapacity = GetCapacity();`

LaserHydra commented on Backpacks.cs:
```diff
@@ -651,6 +661,37 @@ public void Open(BasePlayer looter)
+                    var extraItems = _itemContainer.itemList
+                        .OrderBy(item => item.position)
+                        .Where(item => item.position >= allowedCapacity)
+                        .ToList();
```
I think this doesn't necessarily need to be a list, please remove it unless you disagree.

LaserHydra commented on Backpacks.cs:
```diff
@@ -651,6 +661,37 @@ public void Open(BasePlayer looter)
+                if (looter.userID == OwnerId)
+                {
+                    // Check for items beyond the backpack's capacity
+                    var allowedCapacity = GetCapacity();
```
Possibly compare this to the actual capacity and only continue if the actual capacity it larger than it should be. Or, alternatively, which would likely be a bit more performant, keep a boolean `hasOverflowingItems` variable which is set in `Initialize()` and then checked for and reset here.

LaserHydra reviewed (changes requested):
Thank you for your continued contribution.
Good consideration to resize the backpack first, so that they items don't drop if an admin looks into the backpack.

I found a couple of concerns which I've commented on.

WheteThunger committed:
Performance improvements for overflowing backpack detection

WheteThunger commented:
All comments addressed. The functional differences are:
- Performance was improved. Opening a backpack that is not overflowing should now be just as fast as before
- When an admin views an overflowing backpack, they can now see the overflowing items since the backpack temporarily has max capacity. Previously, the overflowing items would be invisible but still present. With this new behavior, it was necessary to make an additional change to close admin looters when handling overflow since otherwise the admin can potentially see slots that have been removed

WheteThunger committed:
Update README.md localization strings

LaserHydra commented on Backpacks.cs:
```diff
@@ -569,19 +572,28 @@ public void Initialize()
                     _itemContainer.GiveUID();
+                    _itemContainer.capacity = allowedCapacity;
+
+                    if (_itemDataCollection.Count > allowedCapacity)
```
One possibility which doesn't seem covered right now is that they don't have many items, but for some reason they're stored in the last few slots which wouldn't be available anymore.

So instead of checking the count of items, I'd maybe check if there is any item which is supposed to be in a slot which is no longer available.
Probably best to write a method to check that, for separation & reusability.

WheteThunger replied:
That was a case I continuously tested while developing. Testing again confirms it works. For example, when the capacity is 12 (2 rows of 6), I put in 5 items, some on the first row, some one the second. Then I change permission to reduce the capacity to 6 and re-open the backpack. The items on the second row are automatically moved to the first row in the same order.

I haven't worked on this for several weeks so I couldn't remember why this works, and it's not obvious from looking at the code. After some testing, it turns out this is handled by the following branch.

```
else if (_itemContainer.capacity > allowedCapacity)
{
    // The capacity will be reduced later when the backpack is opened by the owner
    hasOverflowingItems = true;
}
```

This actually causes the `hasOverflowingItems` flag to be set every time the capacity is reduced by a permission change, so the `HandleOverflowingItems()` method will be called on the next `Open()` from the owner even if there is no capacity issue. This is probably fine and could be clarified with some renames. However, I thought it through and tested it some more, and I found one issue, which is that if an admin views the backpack, the non-overflowing items that are beyond the capacity may be hidden. I will try to fix this in the next revision.

LaserHydra commented on Backpacks.cs:
```diff
@@ -652,21 +674,84 @@ public void Open(BasePlayer looter)
+            private void HandleOverflowingItems(BasePlayer receiver)
             {
-                foreach (BasePlayer looter in _looters.ToArray())
+                // Close for other looters since we are going to alter the capacity
```
Maybe we can re-open the backpack for those looters after altering the capacity?

WheteThunger replied:
I suppose we could, but that case would happen so rarely, and the impact is so negligible that I don't think it's worth the extra lines of code.

LaserHydra replied:
Fair enough, I just thought it'd be confusing the the admin but yes, the chance of this happening is quite slim.

LaserHydra commented on Backpacks.cs:
```diff
@@ -526,6 +527,8 @@ private class Backpack
             private ItemContainer _itemContainer = new ItemContainer();
             private List<BasePlayer> _looters = new List<BasePlayer>();
 
+            public bool hasOverflowingItems = false;
```
This should likely be private and respect the private field naming convention.

WheteThunger replied:
Will update.

LaserHydra reviewed (changes requested):
Looks good, thanks again.
Just a few small things I've noticed still, but we're almost there!

WheteThunger committed:
Improve handling of overflowing items

- Fix issue where an admin viewing a possibly overflowing backpack would not necessarily be shown all of the items.
- Use the BasePlayer.GiveItem() method so we don't have to manually drop any remaining items.
- Rename hasOverflowingItems field to _hasPossibleOverflow.
- Rename HandleOverflowingItems() method to HandlePossiblyOverflowingItems().

The renames helps to clarify that the Backpack Init() method doesn't always know if there is actual overflow. That uncertainty is resolved when the owner opens their backpack.

LaserHydra commented:
Looks good to me now. Thanks for all your work!
